central/hash/manager/deduper.go

```diff
@@ -0,0 +1,247 @@
+	eventMsg, ok := msg.Msg.(*central.MsgFromSensor_Event)
+	if !ok {
+		return true
+	}
+	if eventMsg.Event.GetProcessIndicator() != nil {
+		return true
+	}
+	// Autogenerated image integrations currently do not have consistent ID
+	if eventMsg.Event.GetImageIntegration() != nil {
+		return true
+	}
+	if alert.IsRuntimeAlertResult(msg.GetEvent().GetAlertResults()) {
+		return true
+	}
+	// This can occur for a very short-lived deployment where alerts are not generated
+	// but the deployment is being removed
+	if alert.IsAlertResultResolved(msg.GetEvent().GetAlertResults()) {
+		return true
+	}
+	if eventMsg.Event.GetReprocessDeployment() != nil {
+		return true
+	}
+	return false
+}
+
+func (d *deduperImpl) shouldReprocess(hashKey string, hash uint64) bool {
+	prevValue, ok := d.getValueNoLock(hashKey)
+	if !ok {
+		// This implies that a REMOVE event has been processed before this event
+		// Note: we may want to handle alerts specifically because we should insert them as already resolved for completeness
+		return false
+	}
+	// This implies that no new event was processed after the initial processing of the current message
+	return prevValue == hash
+}
+
+func getIDFromKey(key string) string {
+	return stringutils.GetAfter(key, ":")
+}
+
+func buildKey(typ, id string) string {
+	return fmt.Sprintf("%s:%s", typ, id)
+}
+
+func getKey(msg *central.MsgFromSensor) string {
+	event := msg.GetEvent()
+	return buildKey(reflectutils.Type(event.GetResource()), event.GetId())
+}
+
+// MarkSuccessful marks a message as successfully processed
+func (d *deduperImpl) MarkSuccessful(msg *central.MsgFromSensor) {
+	// If the object isn't eligible for deduping then do not mark it as being successfully processed
+	// because it does not exist in the received map
+	if skipDedupe(msg) {
+		return
+	}
+	key := getKey(msg)
+	// If we are removing, then we do not need to mark it as successful as there is nothing more
+	// to potentially dedupe. We do need to remove it from successfully processed as ShouldProcess is
+	// evaluated as objects come into the queue and an object may have been successfully processed after
+	if msg.GetEvent().GetAction() == central.ResourceAction_REMOVE_RESOURCE {
+		concurrency.WithLock(&d.hashLock, func() {
+			delete(d.successfullyProcessed, key)
+		})
+		return
+	}
+
+	d.hashLock.Lock()
+	defer d.hashLock.Unlock()
+
+	val, ok := d.received[key]
+	// Only remove from this map if the hash matches as received could contain a more recent event
+	if ok && val.val == msg.GetEvent().GetSensorHash() {
+		delete(d.received, key)
+	}
+	d.successfullyProcessed[key] = &entry{
+		val:       msg.GetEvent().GetSensorHash(),
+		processed: true,
+	}
+}
+
+func (d *deduperImpl) getValueNoLock(key string) (uint64, bool) {
+	if prevValue, ok := d.received[key]; ok {
+		return prevValue.val, ok
+	}
+	prevValue, ok := d.successfullyProcessed[key]
+	if !ok {
+		return 0, false
+	}
+	return prevValue.val, true
+}
+
+func (d *deduperImpl) ProcessSync() {
```

Comment:
This function will only be called once per sensor connection (e.g. if the connection dies, it will be called again once again). So Central could still be alive and receive multiple syncs. Will think about your suggestion. I do like the copy instead of the tracking

Reply:
Okay so after some further thought, I'd prefer to leave it as is:
1. Map clones are still iterations so we're not saving too much there
2. I feel like introducing a new map makes it even more confusing
3. The removal of invalidated resources would still need to take into account the alerts as they are expensive so would be great to continue to dedupe them.

I ran a test of iteration when the size of the map was 1 million (the current default max size) and it was 16ms so in the grand scheme of things, not too consequential